examples/plugins/traffic-control/main.go

```diff
@@ -0,0 +1,234 @@
+func (tcs *TrafficControlStatus) SetLatency(latency string) {
+	if latency == "" {
+		tcs.latency = "-"
+	}
+	tcs.latency = latency
+}
+
+// SetPacketLoss sets the packet loss value
+// the convention is that empty packet loss is represented by '-'
+func (tcs *TrafficControlStatus) SetPacketLoss(packetLoss string) {
+	if packetLoss == "" {
+		tcs.packetLoss = "-"
+	}
+	tcs.packetLoss = packetLoss
+}
+
+// TrafficControlStatusInit initializes with the convention that empty values are '-'
+func TrafficControlStatusInit() *TrafficControlStatus {
+	return &TrafficControlStatus{
+		latency:    "-",
+		packetLoss: "-",
+	}
+}
+
+// TrafficControlStatusCache implements status caching
+var trafficControlStatusCache map[string]*TrafficControlStatus
+
+func main() {
+	const socket = "/var/run/scope/plugins/traffic-control.sock"
+
+	// Handle the exit signal
+	setupSignals(socket)
+
+	listener, err := setupSocket(socket)
+	if err != nil {
+		log.Fatalf("Failed to setup socket: %v", err)
+	}
+
+	plugin, err := NewPlugin()
+	if err != nil {
+		log.Fatalf("Failed to create a plugin: %v", err)
+	}
+
+	// Cache
+	trafficControlStatusCache = make(map[string]*TrafficControlStatus)
+
+	trafficControlServeMux := http.NewServeMux()
+
+	// Report request handler
+	reportHandler := http.HandlerFunc(plugin.report)
+	trafficControlServeMux.Handle("/report", reportHandler)
+
+	// Control request handler
+	controlHandler := http.HandlerFunc(plugin.control)
+	trafficControlServeMux.Handle("/control", controlHandler)
+
+	log.Println("Listening...")
+	if err = http.Serve(listener, trafficControlServeMux); err != nil {
+		log.Fatalf("failed to serve: %v", err)
+	}
+}
+
+func setupSignals(socket string) {
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt)
+	go func() {
+		<-interrupt
+		os.Remove(socket)
+		os.Exit(0)
+	}()
+}
+
+func setupSocket(socket string) (net.Listener, error) {
+	os.Remove(socket)
+	if err := os.MkdirAll(filepath.Dir(socket), 0755); err != nil {
+		return nil, fmt.Errorf("failed to create directory %q: %v", filepath.Dir(socket), err)
+	}
+	listener, err := net.Listen("unix", socket)
+	if err != nil {
+		return nil, fmt.Errorf("failed to listen on %q: %v", socket, err)
+	}
+
+	log.Printf("Listening on: unix://%s", socket)
+	return listener, nil
+}
+
+// NewPlugin ...
+func NewPlugin() (*Plugin, error) {
+	store := NewStore()
+	dockerClient, err := NewDockerClient(store)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a docker client: %v", err)
+	}
+	reporter := NewReporter(store)
+	plugin := &Plugin{
+		reporter: reporter,
+		clients: []containerClient{
+			dockerClient,
+		},
+	}
+	for _, client := range plugin.clients {
+		go client.Start()
+	}
+	return plugin, nil
+}
+
+// Serve ...
+func (p *Plugin) Serve(listener net.Listener) error {
+	http.HandleFunc("/report", p.report)
+	http.HandleFunc("/control", p.control)
+	return http.Serve(listener, nil)
```

Comment:
ok

On Fri, Sep 9, 2016 at 12:00 PM, Paul Bellamy <notifications@github.com>
wrote:

> In examples/plugins/traffic-control/main.go
> <https://github.com/weaveworks/scope/pull/1837#discussion_r78156185>:
>
> > +		reporter: reporter,
> > +		clients: []containerClient{
> > +			dockerClient,
> > +		},
> > +	}
> > +	for _, client := range plugin.clients {
> > +		go client.Start()
> > +	}
> > +	return plugin, nil
> > +}
> > +
> > +// Serve ...
> > +func (p *Plugin) Serve(listener net.Listener) error {
> > +	http.HandleFunc("/report", p.report)
> > +	http.HandleFunc("/control", p.control)
> > +	return http.Serve(listener, nil)
>
> Ah, either way. Doing it here means that Plugin.Serve(net.Listener) error
> has the same type signature as http.ServeMux.Serve, which is kind of nice.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/weaveworks/scope/pull/1837/files/c03aa1cfc25aca328704f1da0674a571f69da134#r78156185>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGRc7x_J2d8d3mFBsDwgRPPVBaCzZi0nks5qoS4ggaJpZM4JyDW_>
> .
>



-- 

Alessandro Puccetti